### PR TITLE
Added transformation, e-shadow, e-gradient parameter

### PR DESCRIPTION
**Pull request by ankur-dwivedi**

### CODE REVIEW

File: test/upload.js [hunk 1]
```diff
@@ -1091,4 +1091,35 @@ describe("File upload", function () {
...
     });
+
+    it('With pre and post transformation', async function () {
+        const fileOptions = {
+            ...securityParameters,
+            fileName: "test_file_name",
+            file: "test_file",
+            responseFields: "tags, customCoordinates, isPrivateFile, metadata",
+            useUniqueFileName: false,
+            transformation: { pre: 'w-100', post: [{type: 'transformation', value: ''}]}
```
Comment (imagekitio) [inline test/upload.js:1102]: It should result in error. why allow empty `value`?

File: test/url-generation.js [hunk 2]
```diff
@@ -263,11 +250,11 @@ describe("URL generation", function () {
         const url = imagekit.url({
             path: "/test_path.jpg",
             transformation: [{
-                overlayImage: ""
+                effectShadow: ""
```
Comment (imagekitio) [inline test/url-generation.js:253]: It should match with convention of passing `true` if I am not wrong. Check how SDK works with other effect parameters that don't necessarily need a value.

File: src/constants/supportedTransforms.ts [hunk 3]
```diff
@@ -87,141 +87,6 @@ const supportedTransforms: { [key: string]: string } = {
    */
   named: "n",
 
-  /**
-   * @link https://docs.imagekit.io/features/image-transformations/overlay#overlay-image-oi
-   */
-  overlayImage: "oi",
```
[the remark this answers is not in the document]
Comment (ankur-dwivedi) [reply src/constants/supportedTransforms.ts:93]: done

File: test/upload.js (hunk 1 above)
Comment (ankur-dwivedi) [reply test/upload.js]: added checks for this and also added test suits to assert this

File: test/url-generation.js (hunk 2 above)
Comment (ankur-dwivedi) [reply test/url-generation.js]: type definition of `effectShadow ` is the same as other effect parameters. I have removed it from the above test suit and we have a test suit with `true` as its value

File: src/interfaces/UploadOptions.ts [hunk 4]
```diff
@@ -1,3 +1,25 @@
...
+  value?: string;
+}
+
+interface AbsObject {
+  type: "abs";
+  value: string;
+  protocol: "hls" | "dash";
+}
+
+type postTransformation = TransformationObject | GifToVideoOrThumbnailObject | AbsObject;
```
Comment (ahnv) [inline src/interfaces/UploadOptions.ts:17]: Interfaces should be named with proper CamelCasing

File: src/interfaces/UploadOptions.ts [hunk 5]
```diff
@@ -1,3 +1,25 @@
...
+
+interface transformation{
```
Comment (ahnv) [inline src/interfaces/UploadOptions.ts:19]: Same here
```
Interfaces should be named with proper CamelCasing
```

**Review by ahnv — CHANGES_REQUESTED** [review]

File: src/interfaces/UploadOptions.ts (hunk 4 above)
Comment (ankur-dwivedi) [reply src/interfaces/UploadOptions.ts]: done

File: src/interfaces/UploadOptions.ts (hunk 5 above)
Comment (ankur-dwivedi) [reply src/interfaces/UploadOptions.ts]: done

**Review by ahnv — APPROVED** [review]
LGTM

File: src/constants/supportedTransforms.ts [hunk 6]
```diff
@@ -282,6 +147,16 @@ const supportedTransforms: { [key: string]: string } = {
    */
   original: "orig",
 
+  /**
+   * @link https://docs.imagekit.io/features/image-transformations/image-enhancement-and-color-manipulation#shadow-e-shadow
+   */
+  effectShadow: "e-shadow",
+
+  /**
+   * @link https://docs.imagekit.io/features/image-transformations/image-enhancement-and-color-manipulation.md#gradient-e-gradient
```
Comment (imagekitio) [inline src/constants/supportedTransforms.ts:156]: "https://docs.imagekit.io/features/image-transformations/image-enhancement-and-color-manipulation.md#gradient-e-gradient" broken url

File: test/upload.js [hunk 7]
```diff
@@ -1091,4 +1091,215 @@ describe("File upload", function () {
...
+      expect(arg.get("fileName")).to.be.equal("test_file_name");
+      expect(arg.get("responseFields")).to.be.equal("tags, customCoordinates, isPrivateFile, metadata");
+      expect(arg.get("useUniqueFileName")).to.be.equal("false");
+      expect(arg.get("publicKey")).to.be.equal("test_public_key");
+      expect(arg.get("transformation")).to.be.equal(JSON.stringify(fileOptions.transformation));
+
+      expect(callback.calledOnce).to.be.true;
+      sinon.assert.calledWith(callback, null, uploadSuccessResponseObj);
+    });
+    it("With pre transformation", async function () {
```
Comment (imagekitio) [inline test/upload.js:1125]: New line missing.

File: test/upload.js [hunk 8]
```diff
@@ -1091,4 +1091,215 @@ describe("File upload", function () {
...
+      };
+      var callback = sinon.spy();
+
+      imagekit.upload(fileOptions, callback);
+
+      expect(server.requests.length).to.be.equal(1);
+      await sleep();
+      var errRes = {
+        help: "",
+        message: "Invalid transformation parameter.",
```
Comment (imagekitio) [inline test/upload.js:1202]: Error message is not helpful to user. We can add a valid example like we do in our transformation parser.

File: test/upload.js (hunk 7 above)
Comment (ankur-dwivedi) [reply test/upload.js:1126]: done

File: src/constants/supportedTransforms.ts (hunk 6 above)
Comment (ankur-dwivedi) [reply src/constants/supportedTransforms.ts]: done

File: test/upload.js (hunk 8 above)
Comment (ankur-dwivedi) [reply test/upload.js]: done

**Pull request merged**